packages/Adapters/POD_C/src/DTK_C_API.cpp

```diff
@@ -0,0 +1,246 @@
+#include <boost/property_tree/ptree.hpp>
+#include <boost/property_tree/info_parser.hpp>
+
+#include <sstream> 
+#include <iostream>
+#include <string>
+
+//----------------------------------------------------------------------------//
+Teuchos::RCP<Tpetra::Map<int,DataTransferKit::EntityId> const>
+  build_contiguous_map(
+    Teuchos::RCP<Teuchos::Comm<int> const> const& comm,
+    DataTransferKit::EntityId local_num )
+{
+  DataTransferKit::EntityId global_num = 0;
+  Teuchos::reduceAll<int,DataTransferKit::EntityId>( *comm,
+                                                     Teuchos::REDUCE_SUM,
+                                                     1,
+                                                     &local_num,
+                                                     &global_num );
+  return Tpetra::createContigMap<int, DataTransferKit::EntityId>(
+    global_num, local_num, comm );
+}
+
+//----------------------------------------------------------------------------//
+Teuchos::Array<DataTransferKit::Entity>
+  pull_entities(
+    Teuchos::RCP<Teuchos::Comm<int> const> const& comm,
+    double const* points,
+    unsigned local_num,
+    DTK_Data_layout data_layout,
+    int space_dim )
+{
+  Teuchos::Array<DataTransferKit::Entity> entities(local_num);
+  Teuchos::Array<double> coord(space_dim);
+  DataTransferKit::EntityId global_id;
+  int const comm_size = comm->getSize();
+  int const comm_rank = comm->getRank();
+
+  // @Stuart: any way to do that with some teuchos rountine?
+  // I guess the tpetra map can give this information
+  // Then I'd pass the map instead of comm and local_num
+  std::vector<unsigned> shift(comm_size);
+  Teuchos::gatherAll<int,unsigned>( *comm,
+                                    1,
+                                    &local_num,
+                                    shift.size(),
+                                    shift.data() );
+  for (std::size_t k = 1; k < comm_size; ++k)
+    shift[k] = shift[k-1];
+  shift[0] = 0;
+  for (std::size_t k = 1; k < comm_size; ++k)
+    shift[k] += shift[k-1];
+
+  // TODO: might be worth implementing some data extractor to be used both here
+  // and when handling fields
+  if (data_layout == DTK_INTERLEAVE)
+    for (unsigned i = 0; i < local_num; ++i)
+    {
+      global_id = shift[comm_rank] + i;
+      for (int j = 0; j < space_dim; ++j)
+        coord[j] = points[space_dim*i+j];
+      entities[i] = DataTransferKit::Point(global_id, comm_rank, coord);
+    }
+  else if (data_layout == DTK_CONTIGUOUS)
+    for (unsigned i = 0; i < local_num; ++i)
+    {
+      global_id = shift[comm_rank] + i;
+      for (int j = 0; j < space_dim; ++j)
+        coord[j] = points[i+local_num*j];
+      entities[i] = DataTransferKit::Point(global_id, comm_rank, coord);
+    }
+  else
+    throw std::runtime_error("Invalid data layout");
+  return entities;
+}
+
+//----------------------------------------------------------------------------//
+Teuchos::Array<double>
+  pull_field(
```

Comment:
Entity centered field will handle both pull_field and push_field for interleaved and blocked data